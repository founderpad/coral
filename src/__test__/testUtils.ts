// import { ThemeProvider } from "my-ui-lib"
// import { TranslationProvider } from "my-i18n-lib"
// import defaultStrings from "i18n/en-x-default"
import { render } from '@testing-library/react';
import { JSXElementConstructor, ReactElement } from 'react';

const Providers = ({ children }: { children: any }) => {
	return children;
	// return (
	//   <ThemeProvider theme="light">
	//     <TranslationProvider messages={defaultStrings}>
	//       {children}
	//     </TranslationProvider>
	//   </ThemeProvider>
	// )
};

const customRender = (
	ui: ReactElement<any, string | JSXElementConstructor<any>>,
	options = {}
): any => render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
