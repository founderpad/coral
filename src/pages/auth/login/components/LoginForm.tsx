import { AlertFeedback } from '@components/alert';
import { SubmitButton } from '@components/buttons';
import { BaseForm } from '@components/form';
import { FormInput } from '@components/form/inputs/FormField';
import { Label } from '@components/labels';
import { FlexLayout } from '@components/layouts';
import { PrimaryLink } from '@components/links';
import { useLogin } from '@hooks/auth';
import { useQueryParam } from '@hooks/util';
import { emailPattern } from '@utils/validators';
import React, { memo } from 'react';
import { TLoginFields } from '../../../../types/auth';

const defaultValues = {
	email: '',
	password: ''
};

const LoginForm = () => {
	const onLogin = useLogin();
	const isError = useQueryParam('error');
	const isVerifiedEmail = useQueryParam('type') === 'emailVerify';

	return (
		<React.Fragment>
			{isVerifiedEmail && (
				<AlertFeedback message="Email address successfully verified" />
			)}

			<BaseForm<TLoginFields>
				name="login-form"
				onSubmit={onLogin}
				defaultValues={defaultValues}
				stackProps={{
					alignItems: 'center',
					spacing: 2
				}}
			>
				{({
					register,
					control,
					resetField,
					formState: { errors, isSubmitting }
				}) => (
					<React.Fragment>
						<FormInput<TLoginFields>
							id="email"
							name="email"
							placeholder="Email"
							register={register}
							control={control}
							fieldProps={{
								placeholder: 'Email'
							}}
							rules={{
								required:
									'You must enter a valid email address',
								pattern: emailPattern
							}}
							hideLimit={true}
							errors={errors}
							onClear={() => resetField('email')}
						/>

						<FormInput<TLoginFields>
							id="password"
							name="password"
							placeholder="Password"
							register={register}
							control={control}
							fieldProps={{
								placeholder: 'Password',
								type: 'password'
							}}
							rules={{
								required: 'You must enter a valid password',
								minLength: {
									value: 6,
									message:
										'Your password must be a minimum of 6 characters'
								},
								maxLength: {
									value: 20,
									message:
										'Your password must be a maximum of 20 characters'
								}
							}}
							hideLimit={true}
							errors={errors}
							onClear={() => resetField('password')}
						/>

						{isError && (
							<AlertFeedback
								status="error"
								message={
									'Failed to login. Incorrect email and/or password.'
								}
							/>
						)}

						<SubmitButton
							id="submit-login"
							name="submit-login"
							label="Log in"
							isLoading={isSubmitting}
							disabled={isSubmitting}
							size="md"
							fontSize="small"
							w={{ base: 'full', sm: '150px' }}
						/>
					</React.Fragment>
				)}
			</BaseForm>
			{/* <SocialLogins /> */}
			<LoginFooter />
		</React.Fragment>
	);
};

const LoginFooter = memo(() => (
	<FlexLayout justifyContent="space-between" alignItems="center" pt={8}>
		<Label color="gray.500" fontSize="x-small" alignSelf="center">
			No account?{' '}
			<PrimaryLink href="/register" title="Link to register an account">
				Register now
			</PrimaryLink>
		</Label>
		<PrimaryLink
			href="/resetpassword"
			title="Link to register an account"
			fontSize="x-small"
		>
			Forgotten password?
		</PrimaryLink>
	</FlexLayout>
));

export default LoginForm;
