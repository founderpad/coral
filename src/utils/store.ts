import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import { persistReducer } from 'redux-persist';
import { PersistConfig } from 'redux-persist/es/types';
import storage from 'redux-persist/lib/storage';
import rootReducer from './reducer';

/**
 * Create the Redux store
 *
 * @author jlee
 */

// const isProd = process.env.NODE_ENV !== 'development';

/**
 * Persist storage configuration set up
 * @key[root]
 * @storage[storage]
 */
const persistConfig: PersistConfig<any> = {
	key: 'root',
	storage
};

const middleware = [...getDefaultMiddleware({ serializableCheck: false })];

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
	reducer: persistedReducer,
	middleware
	// devTools: isProd ?? false
});

// export { StoreProvider }

export default store;
