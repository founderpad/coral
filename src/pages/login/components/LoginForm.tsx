import { Flex, Text } from '@chakra-ui/layout';
import { SubmitButton } from 'components/buttons';
import { Form } from 'components/form';
import { useBaseForm } from 'components/form/hooks';
import { EmailField, PasswordField } from 'components/input';
import { StackLayout } from 'components/layouts';
import { PrimaryLink } from 'components/links';
import { useLogin } from 'hooks/auth';
import React, { memo } from 'react';
import { IAuthFormData } from 'types/auth';

const LoginForm = (): JSX.Element => {
	const {
		handleSubmit,
		control,
		formState: { errors, isSubmitting, isValid }
	} = useBaseForm<IAuthFormData>();
	const onLogin = useLogin();

	return (
		<Form
			id={'login-form'}
			name={'loginform'}
			onSubmit={handleSubmit(onLogin)}
		>
			<StackLayout spacing={6} my={6}>
				<EmailField
					id="email"
					name="email"
					error={errors['email']}
					errorText="Please enter a valid email"
					control={control}
					isRequired
				/>
				<PasswordField
					id="password"
					name="password"
					error={errors['password']}
					rules={{ maxLength: 20, minLength: 6 }}
					control={control}
					isRequired
				/>

				<SubmitButton
					id={'submit-login'}
					name={'submit-login'}
					label="Login now"
					isLoading={isSubmitting}
					disabled={!isValid || isSubmitting}
					size={'md'}
					full
				/>
				<NoAccountFooter />
			</StackLayout>
		</Form>
	);
};

const NoAccountFooter = memo(
	(): JSX.Element => (
		<Flex
			spacing={6}
			justifyContent={'center'}
			w="full"
			mt={'auto'}
			mx={'auto'}
		>
			<Text color={'gray.500'} fontSize={'xs'}>
				Don&apos;t have an account?
			</Text>
			&nbsp;
			<PrimaryLink
				href="/register"
				fontSize={'xs'}
				title={'Link to register an account'}
			>
				Create one here
			</PrimaryLink>
		</Flex>
	)
);

export default LoginForm;
