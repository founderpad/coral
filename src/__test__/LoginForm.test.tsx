import LoginForm from '@pages/login/components/LoginForm';
import userEvent from '@testing-library/user-event';
import store from '@utils/store';
import { Provider } from 'react-redux';
import { act, cleanup, fireEvent, render, waitFor } from './testUtils';

const setup = () => {
	const loginSetup = render(
		<Provider store={store}>
			<LoginForm />
		</Provider>
	);

	const loginForm = loginSetup.getByRole('form', { name: /loginform/i });
	const emailField = loginSetup.getByRole('textbox', { name: /email/i });
	const passwordField = loginSetup.getByPlaceholderText(/Password/i);
	const submitButton = loginSetup.getByRole('button', { name: /submit/i });

	return {
		loginSetup,
		loginForm,
		emailField,
		passwordField,
		submitButton,
		...loginSetup
	};
};

const mockLogin = jest.fn();
jest.mock('@hooks/auth', () => ({
	useLogin: (): any => mockLogin
}));

describe('Login form', () => {
	afterEach(cleanup);
	it('matches snapshot', async () => {
		const { asFragment } = setup();

		await waitFor(() => asFragment());
		expect(asFragment()).toMatchSnapshot();
	});

	it('should render LoginForm', async () => {
		const { loginForm } = setup();
		expect(loginForm).toBeInTheDocument();
	});

	it('should render LoginForm fields', async () => {
		const { submitButton } = setup();
		expect(submitButton).toBeDisabled();
	});

	it('should make login request on submit', async () => {
		const { emailField, passwordField, submitButton } = setup();

		expect(emailField).toBeInTheDocument();
		expect(passwordField).toBeInTheDocument();

		userEvent.type(emailField, 'jamie@gmail.com');
		await waitFor(() => expect(emailField).toHaveValue('jamie@gmail.com'));

		userEvent.type(passwordField, 'admin123456');
		await waitFor(() => expect(passwordField).toHaveValue('admin123456'));

		await act(async () => {
			fireEvent.click(submitButton);
		});

		expect(mockLogin).toHaveBeenCalledTimes(1);
	});

	it('should error on empty email on blur', async () => {
		const { loginSetup, emailField, passwordField } = setup();

		expect(emailField).toBeInTheDocument();
		expect(passwordField).toBeInTheDocument();

		fireEvent.focus(emailField);
		userEvent.type(emailField, 'j@gmail');
		await waitFor(() => expect(emailField).toHaveValue('j@gmail'));

		fireEvent.blur(emailField);

		loginSetup.getByText('Please enter a valid email');
	});

	it('should error on empty password on blur', async () => {
		const { loginSetup, emailField, passwordField } = setup();

		expect(emailField).toBeInTheDocument();
		expect(passwordField).toBeInTheDocument();

		fireEvent.focus(passwordField);
		userEvent.type(passwordField, 'pwd');
		await waitFor(() => expect(passwordField).toHaveValue('pwd'));

		fireEvent.blur(passwordField);

		loginSetup.getByText(
			'Please enter a valid password between 6 and 20 characters'
		);
	});
});
