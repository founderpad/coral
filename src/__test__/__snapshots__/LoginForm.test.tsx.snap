// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login form matches snapshot 1`] = `
<DocumentFragment>
  <form
    aria-label="login-form"
    autocomplete="off"
    id="login-form"
    name="login-form"
    novalidate=""
  >
    <div
      class="chakra-stack css-ew27cg"
    >
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <div
          aria-live="polite"
          class="css-jhkd4z"
          id="form-input"
        >
          <div
            class="css-qnoc8j"
          />
          <input
            aria-label="email"
            aria-required="true"
            class="chakra-input css-gr2y0"
            id="field-1"
            placeholder="Email"
            required=""
            value=""
          />
          <div
            class="css-11633wt"
          />
        </div>
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <div
          aria-live="polite"
          class="css-jhkd4z"
          id="form-input"
        >
          <div
            class="css-qnoc8j"
          />
          <input
            aria-label="password"
            aria-required="true"
            class="chakra-input css-gr2y0"
            id="field-2"
            placeholder="Password"
            required=""
            type="password"
            value=""
          />
          <div
            class="css-11633wt"
          />
        </div>
      </div>
      <div
        class="chakra-alert css-glexlw"
        role="alert"
      >
        <svg
          class="chakra-alert__icon css-1tw3v2n"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M368 368L144 144m224 0L144 368"
            fill="none"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="32"
          />
        </svg>
        <div
          class="chakra-alert__desc css-1g4x3gd"
        >
          Failed to login. Incorrect email and/or password.
        </div>
      </div>
      <button
        aria-label="submit-login"
        class="chakra-button css-1hm0q50"
        id="submit-login"
        type="submit"
      >
        Log in
      </button>
    </div>
  </form>
  <div
    class="css-18voih5"
  >
    <p
      class="chakra-text css-1fihqrz"
    >
      No account? 
      <span
        class="chakra-link css-ntv2za"
        href="/register"
        title="Link to register an account"
      >
        Register now
      </span>
    </p>
    <span
      class="chakra-link css-pkysmm"
      href="/resetpassword"
      title="Link to register an account"
    >
      Forgotten password?
    </span>
  </div>
</DocumentFragment>
`;
